### DISCUSSION TITLE
Disable and (re-)enable Sentry after Startup in .NET MAUI

### DISCUSSION THREAD

**Discussion by FlavioGoncalves-Cayas** [getsentry/sentry-dotnet#3325] [Q&A]
Hi folks,

due to data privacy concerns we want to give the use the option to turn off sentry at runtime.
As far as I understand there are generally three options to do that:
1. Dispose `IDisposable` returned during `Init`
2. Call `SentrySdk.Close()`
3. Initialize Sentry with empty DSN

In .NET MAUI option 1 is not feasible when using the recommended approach to configure Sentry with `UseSentry()`.

Option 2 can be done effortlessly, but the problem I see is when we want to reenable Sentry. Then we will have to call `Init()`. However, calling `Init()` manually or `UseSentry()` is not the same in a MAUI app, because `UseSentry()` does more than just initialize Sentry. It also registers event handlers for lifecycle, navigation, page events and more, so that they can be reported to Sentry. Since all of that logic is internal it is also not possible to do that outside of `UseSentry()`.

Third option also requires a new call to `Init()`.

I also tried implementing the `SetBeforeSend` and `SetBeforeSendTransaction` and drop all events in case the user disabled sentry, but even then crashes, ANR events and general information like the information found in Release Details seems to still be transmitted, which is not desirable.

Am I missing an obvious API to turn all data transmission off, or is this an oversight especially in regards to the MAUI integration?

@bitsandfoxes @jamescrosswell mentioning you for visibility.

**Comment by jamescrosswell** [comment]
I can't see a way to do this currently. 

Brainstorming on how this _might_ be done, is there any way to [Configure Lifecycle Events](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.lifecycleevents.mauiapphostbuilderextensions.configurelifecycleevents?view=net-maui-8.0&viewFallbackFrom=net-7.0) for MAUI outside of the `MauiAppBuilder`?

**Reply by FlavioGoncalves-Cayas** [reply]
In the current version of MAUI (8.0.20) this could be done in the following way:
```
(MauiUIApplicationDelegate.Current.Services.GetService<ILifecycleEventService>() as ILifecycleBuilder)?.AddiOS(lcb =>
{
    lcb.DidEnterBackground(application =>
    {
            
    });
});
```
However, as you can see, one would need to cast ILifecycleEventService to ILifecycleBuilder, since the extension method `AddiOS` is defined for interface `ILifecycleEventBuilder`. This could be considered as a hack and only works because [LifecycleEventService](https://github.com/dotnet/maui/blob/main/src/Core/src/LifecycleEvents/LifecycleEventService.cs) implements both interfaces. Refer to [this](https://github.com/davidortinau/WeatherTwentyOne/blob/main/src/WeatherTwentyOne/Services/ServiceExtensions.cs) for access to the `ServiceProvider`.

The advantage of this approach would be, that the same code from [RegisterMauiEventsBinder()](https://github.com/getsentry/sentry-dotnet/blob/main/src/Sentry.Maui/SentryMauiAppBuilderExtensions.cs#L64) can be reused.

Of course, there are still other (platformspecifc) ways to attach to the LifecycleEvents.

**Reply by jamescrosswell** [reply]
Hm, sounds like it _might_ be possible... 

That would be coding up something just for MAUI. Something similar would then have to be done to add "re-enable" functionality for all the other integrations as well (e.g. ASP.NET Core, UWP etc.). 

I wonder if there's a more generic way to be able to dynamically disable/re-enable the SDK.

In the short term, it probably requires either restarting the application or using [Filtering](https://docs.sentry.io/platforms/dotnet/configuration/filtering/).

**Reply by FlavioGoncalves-Cayas** [reply]
> I wonder if there's a more generic way to be able to dynamically disable/re-enable the SDK.

Normally disable/re-enable can be done with Close/Dispose and Init. It's only the integrations that do more than just calling Init (which generally is a nice thing).

> In the short term, it probably requires either restarting the application 

Yes, restarting helps. It's only the current application session, that behaves differently after re-enabling.

> or using [Filtering](https://docs.sentry.io/platforms/dotnet/configuration/filtering/).

When I tried to suppress all events with filtering, native ANRs were still coming through.

**Reply by FlavioGoncalves-Cayas** [reply]
Also noticed that when Sentry was enabled via `UseSentry` then disabled and re-enabled via `Init`, then Sentry will still add breadcrumbs for lifecycle events (I think) via the Hub Instance configured previously in `UseSentry`. That is because `Close/Dispose` does not remove the event handlers.

Might be an issue, if sentry is also configured differently during `Init` for example with another DSN.
